internal/jbig2/segments/generic-refinement-region.go

```diff
@@ -0,0 +1,1020 @@
+				if err != nil {
+					return err
+				}
+				c3 |= int16(temp<<2) & 0x04
+			}
+		}
+
+		if shiftOffset == 0 {
+			w1 = 0
+			w2 = 0
+			w3 = 0
+
+			if modRefByteIdx < refRowStride-1 {
+				if currentLine >= 1 && currentLine-1 < g.ReferenceBitmap.Height {
+					temp, err = g.ReferenceBitmap.GetByte(refByteIndex - refRowStride)
+					if err != nil {
+						return err
+					}
+					w1 = int(temp)
+				}
+
+				if currentLine >= 0 && currentLine < g.ReferenceBitmap.Height {
+					temp, err = g.ReferenceBitmap.GetByte(refByteIndex)
+					if err != nil {
+						return err
+					}
+					w2 = int(temp)
+				}
+
+				if currentLine >= -1 && currentLine+1 < g.ReferenceBitmap.Height {
+					temp, err = g.ReferenceBitmap.GetByte(refByteIndex + refRowStride)
+					if err != nil {
+						return err
+					}
+					w3 = int(temp)
+				}
+			}
+			refByteIndex++
+		}
+	} else {
+		c1 = int16(w1<<1) & 0x07
+		c2 = int16(w2<<1) & 0x07
+		c3 = int16(w3<<1) & 0x07
+		w1 = 0
+		w2 = 0
+		w3 = 0
+
+		if modRefByteIdx < refRowStride-1 {
+			if currentLine >= 1 && currentLine-1 < g.ReferenceBitmap.Height {
+				temp, err = g.ReferenceBitmap.GetByte(refByteIndex - refRowStride)
+				if err != nil {
+					return err
+				}
+				w1 = int(temp)
+			}
+
+			if currentLine >= 0 && currentLine < g.ReferenceBitmap.Height {
+				temp, err = g.ReferenceBitmap.GetByte(refByteIndex)
+				if err != nil {
+					return err
+				}
+				w2 = int(temp)
+			}
+
+			if currentLine >= -1 && currentLine+1 < g.ReferenceBitmap.Height {
+				temp, err = g.ReferenceBitmap.GetByte(refByteIndex + refRowStride)
+				if err != nil {
+					return err
+				}
+				w3 = int(temp)
+			}
+			refByteIndex++
+		}
+
+		c1 |= int16((w1 >> 7) & 0x07)
+		c2 |= int16((w2 >> 7) & 0x07)
+		c3 |= int16((w3 >> 7) & 0x07)
+	}
+	c4 = int16(w4 >> 6)
+	c5 = 0
+
+	modBitsToTrim := (2 - modReferenceDX) % 8
+	w1 <<= uint(modBitsToTrim)
+	w2 <<= uint(modBitsToTrim)
+	w3 <<= uint(modBitsToTrim)
+	w4 <<= 2
+	var bit int
+
+	for x := 0; x < width; x++ {
+		minorX := x & 0x07
+		tval := templateFormation.form(c1, c2, c3, c4, c5)
+
+		if g.override {
+			temp, err = g.RegionBitmap.GetByte(g.RegionBitmap.GetByteIndex(x, lineNumber))
+			if err != nil {
+				return err
+			}
+			g.cx.SetIndex(g.overrideAtTemplate0(int(tval), x, lineNumber, int(temp), minorX))
+		} else {
+			g.cx.SetIndex(int(tval))
+		}
+
+		bit, err = g.arithDecode.DecodeBit(g.cx)
+		if err != nil {
+			return err
+		}
+
+		if err = g.RegionBitmap.SetPixel(x, lineNumber, byte(bit)); err != nil {
+			return err
+		}
+
+		c1 = (((c1 << 1) | 0x01&int16(w1>>7)) & 0x07)
+		c2 = (((c2 << 1) | 0x01&int16(w2>>7)) & 0x07)
+		c3 = (((c3 << 1) | 0x01&int16(w3>>7)) & 0x07)
+		c4 = (((c4 << 1) | 0x01&int16(w4>>7)) & 0x07)
+		c5 = int16(bit)
+
+		if (x-g.ReferenceDX)%8 == 5 {
+			if ((x-g.ReferenceDX)/8)+1 >= g.ReferenceBitmap.RowStride {
+				w1 = 0
+				w2 = 0
+				w3 = 0
+			} else {
+				if currentLine >= 1 && (currentLine-1) < g.ReferenceBitmap.Height {
```

Comment:
Could set `w1`, `w2` and `w3` before each if statement and remove the else branches.